### ISSUE TITLE
using modern web technologies like bootstrap

### ISSUE THREAD

**Issue by mikedevita** [Dan-in-CA/OSPi#93]
Have you ever thought about using modern client-side web technologies like bootstrap and some sort of client-side framework like angular for the sprinkler webui?

It'd clean things up quite a bit and make the code more maintainable. I'd be happy to fork and make an attempt if its something you'd consider merging it in.

**Comment by Dan-in-CA** [comment]
I'm certainly willing to consider it.
Do you think it could be done without breaking existing plugins?

When I started this project the idea was just to port Ray's existing Arduino code to Python so it could run on Raspi. I just wanted an irrigation controller to tinker with. Things have kept growing from there.

I chose web.py to implement the UI partly as a tribute to Aaron Shwartz and I wanted to make the software package something that could be downloaded and run without any other set up.

Could you put together a proof of concept prototype?

**Comment by mikedevita** [comment]
I certainly could. Worst case I can use it as my own version I suppose. Let me get something started. I'll let you know.

**Comment by jonathan-marsh** [comment]
Using such a framework would indeed improve usability (mobile friendly, etc.).

One reason though that I didn't use a framework other than core jQuery (which is pretty much a universal standard at this point) is to limit the number of layers contributors to the project would need to learn to be productive.  Choosing a framework that is minimal in scope and learning curve, and widely understood and adopted, would help maintain that principle.  Would love to know which frameworks you think best qualify!

**Comment by Dan-in-CA** [comment]
This is somewhat off topic but this write up in Ars Technina yesterday has a screen shot of the current UI:
http://arstechnica.com/gadgets/2015/02/how-i-upgraded-my-gardens-ugly-drip-system-with-a-sexy-opensprinkler/

**Comment by Dan-in-CA** [comment]
Another thing to consider is can we still use the current translation files. So far there is Czech, Spanish and French. The Spanish and French will probably continue to be updated but I'm not sure about Czech.

**Comment by mikedevita** [comment]
translation files have nothing really to do with it as you're just changing up how that data is presented I think.

[Bootstrap](http://getbootstrap.com) is a pretty easy curve if you ask me. Its mobile first design, most designers/developers are familiar with it and it has amazing documentation.

Other possibility is  [Zurb Foundation](http://foundation.zurb.com/) which I haven't personally used in production yet, but from what I can tell it offers similar base features. 

Basically what I'd like to see is [bootsrap](http://getbootstrap.com) with [font-awesome](fortawesome.github.io/Font-Awesome/) and jquery. Done right would provide a mobile friendly web-ui thats completely responsive and sexy to boot. Ill see what I can work up in the next few days if I have some downtime.. 

someone else can chime in and get it kickstarted if they like.

**Comment by martinpihrt** [comment]
Hi all,
I suppose Rimco refactor version will (when completed) merged into Dan repository? Or Dan and Rimco create each own version (I do not understand)? I contribute to a newer version refactor, I created numerous plug-ins to the newer version - refactor). Of course I will in the future contribute to the Czech language, but into refactor Rimco version (with my plugins, Rimco instalators, logging etc...). I thank you for your reply and clarification of how it actually is (Dan vs Rimco version).
Martin

**Comment by martinpihrt** [comment]
actual screen for Rimco revision:
![email plugin](https://cloud.githubusercontent.com/assets/7847139/6343866/93072742-bbf0-11e4-9bb6-adbd53ebd2f5.jpg)
![help](https://cloud.githubusercontent.com/assets/7847139/6343864/930662d0-bbf0-11e4-9155-7e9ddaa9cb48.jpg)
![helplcdplug1-2](https://cloud.githubusercontent.com/assets/7847139/6343865/93067478-bbf0-11e4-83ca-39aefc4ec57e.jpg)
![helplcdplug2-2](https://cloud.githubusercontent.com/assets/7847139/6343867/9307b69e-bbf0-11e4-8804-e8ed7ccd064b.jpg)
![home](https://cloud.githubusercontent.com/assets/7847139/6343862/9303e29e-bbf0-11e4-9774-2f32b0ec3a98.jpg)
![lcd plugin](https://cloud.githubusercontent.com/assets/7847139/6343863/9305e3f0-bbf0-11e4-8fe4-7dd07899bb6e.jpg)
![level plugin](https://cloud.githubusercontent.com/assets/7847139/6343868/931b3840-bbf0-11e4-956a-266f6e6bb3ff.jpg)
![log](https://cloud.githubusercontent.com/assets/7847139/6343869/931dc330-bbf0-11e4-9288-73214f4823c4.jpg)
![options](https://cloud.githubusercontent.com/assets/7847139/6343871/931e7a1e-bbf0-11e4-8c3d-b4a0ab4e3c16.jpg)
![plugins](https://cloud.githubusercontent.com/assets/7847139/6343870/931e113c-bbf0-11e4-9777-ebf17a4fda80.jpg)
![pressuremonitor plugin](https://cloud.githubusercontent.com/assets/7847139/6343873/931f1514-bbf0-11e4-8861-2a85129dcb56.jpg)
![programs](https://cloud.githubusercontent.com/assets/7847139/6343872/931ea6b0-bbf0-11e4-9d03-27064be55cf6.jpg)
![programs-new](https://cloud.githubusercontent.com/assets/7847139/6343874/9333a4de-bbf0-11e4-8b04-557be42fb23d.jpg)
![run-once](https://cloud.githubusercontent.com/assets/7847139/6343875/9335c6ec-bbf0-11e4-899d-3934d8900089.jpg)
![sms modem plugin](https://cloud.githubusercontent.com/assets/7847139/6343877/933757d2-bbf0-11e4-9be4-8cf79ceaa18c.jpg)
![stations](https://cloud.githubusercontent.com/assets/7847139/6343879/93385e34-bbf0-11e4-9f46-7f68c5309ead.jpg)
![system debug plugin](https://cloud.githubusercontent.com/assets/7847139/6343876/933727c6-bbf0-11e4-9fe0-0500a1d795ad.jpg)
![system info plugin](https://cloud.githubusercontent.com/assets/7847139/6343878/933782b6-bbf0-11e4-9612-b40af9b7f7d4.jpg)
![update plugin](https://cloud.githubusercontent.com/assets/7847139/6343880/934b2622-bbf0-11e4-9c63-85608df7aac6.jpg)
![upsmonitor plugin](https://cloud.githubusercontent.com/assets/7847139/6343881/934db7ac-bbf0-11e4-84d6-82a2d0b39e63.jpg)
![volttempplugin](https://cloud.githubusercontent.com/assets/7847139/6343883/934e944c-bbf0-11e4-9e6d-2246500cf622.jpg)
![water meter plugin](https://cloud.githubusercontent.com/assets/7847139/6343882/934e5c52-bbf0-11e4-8a39-8051b9c0fe5e.jpg)
![weather delay plugin](https://cloud.githubusercontent.com/assets/7847139/6343884/934f4fd6-bbf0-11e4-880a-649518169a3b.jpg)
![weather level plugin](https://cloud.githubusercontent.com/assets/7847139/6343885/9351450c-bbf0-11e4-8d4f-570f66bba3c5.jpg)
![webcam plugin](https://cloud.githubusercontent.com/assets/7847139/6343886/93624b04-bbf0-11e4-8cbc-76539d613925.jpg)
![wind speed plugin](https://cloud.githubusercontent.com/assets/7847139/6343887/9364bbbe-bbf0-11e4-9d36-7efa543ec455.jpg)
Thanks Martin

**Comment by Dan-in-CA** [comment]
Hi Martin,

Nice to hear from you! 

There are some differences in what the two version of of the project are trying to accomplish. At this point it would be hard to merge Rimco's refactor branch into my original repo.

When Rimco first started his fork he did it in a way that made it difficult for me to contribute to my own project. In fact, the GitHub documentation describes the situation:

```"Historically, the term “fork” has been somewhat negative in context, meaning that someone took an open source project in a different direction, sometimes creating a competing project and splitting the contributors. In GitHub, a “fork” is simply the same project in your own namespace, allowing you to make changes to a project publicly as a way to contribute in a more open manner."```

I have always tried to keep thing simple and accessible so people of every skill level can contribute. It is supposed to be a fun DIY type of project.

There are some features that already exist in my version such as multiple language support, the separate plugins repository, the plugin manager and Blinker signalling which eliminates the need to poll the program for changes. I know Rimco is planning to add some of those to  his version in the future but they probably will done in a different way.

There are also many plugins from several users that have been contributed based on this version.and I have been writing documentation on how to write plugins that would not be valid for his.
https://github.com/Dan-in-CA/ospi_plugins/wiki

Dan

**Comment by martinpihrt** [comment]
Hi Dan,
thank you for your response. I'm sorry, but due to a lot of the time that I spent OSPy fork can not imagine that I would plugins rebuilt back to OSPi version. 
Thank you very much Martin

**Comment by Dan-in-CA** [comment]
Hi Martin,

I understand. Thank you for your previous contributions.
It is because of  your request for translation support in the UI that we 
now have that feature.

Dan

**Comment by mikedevita** [comment]
hey @Dan-in-CA how do you typically do your dev work? Normally on a raspi so you can test sprinkler functions or do you do the web work and the push to a raspi to test? Let me know so I can setup a similar environment for doing the web work.

**Comment by Dan-in-CA** [comment]
Hi Mike,
Up till now I have been doing the dev on a Win 7 system. Lately I have been using PyCharm Pro with an open source license. It can send files to the pi but even that is a bit cumbersome. With the new Pi 2 I have started testing a Pi based dev system but so far it has been a bit flaky.

Do you know of a good (free) Linux app for web dev? I have been mostly focused on Python coding and Jonathan Marsh has been designing the UI. I think we are both open to you input.

**Comment by mikedevita** [comment]
I dev on a macbook mostly. I use Atom text editor, it looks like theres a linux version [here](https://github.com/atom/atom/blob/master/docs/build-instructions/linux.md). Problem though is how do you actually test code without having access to GPIO? 

**Example:**
Lets say you dev away on a win7 box, codes done now you need to make sure stuff works (i am always bouncing back and forth between browser and developing). Do you push your code and the pull it onto the pi and then visit the URL on the Pi? Thats the only way I know of to test. 

Unless of course like you said you dev on the pi itself.

**Comment by mikedevita** [comment]
We should take this offline as its getting off topic. Shoot me an email mike@relative.media.

**Comment by jonathan-marsh** [comment]
For dev code (especially on the UI!) I edit in Eclipse and just run Python
locally on my mac, seems to do everything except actually turning valves on
and off.  Just refresh the browser to test new UIs.  I keep an eye on the
data files to ensure stuff is getting set correctly, and upload to the Pi
to test when I’m ready for a more comprehensive test.



-Jonathan



*From:* Mike DeVita [mailto:notifications@github.com]
*Sent:* Wednesday, February 25, 2015 10:21 AM
*To:* Dan-in-CA/OSPi
*Cc:* Jonathan Marsh
*Subject:* Re: [OSPi] using modern web technologies like bootstrap (#93)



I dev on a macbook mostly. I use Atom text editor, it looks like theres a
linux version here
<https://github.com/atom/atom/blob/master/docs/build-instructions/linux.md>.
Problem though is how do you actually test code without having access to
GPIO?

*Example:*
Lets say you dev away on a win7 box, codes done now you need to make sure
stuff works (i am always bouncing back and forth between browser and
developing). Do you push your code and the pull it onto the pi and then
visit the URL on the Pi? Thats the only way I know of to test.

Unless of course like you said you dev on the pi itself.

—
Reply to this email directly or view it on GitHub
<https://github.com/Dan-in-CA/OSPi/issues/93#issuecomment-76022565>.

**Comment by mikedevita** [comment]
@jonathan-marsh server actually starts up without being able to talk to the gpio's? Wild. I know when I was dev'ing the nodejs version of a sprinkler app it would barf if it couldn't init the GPIO pins.

**Comment by jonathan-marsh** [comment]
I think we put in a bit of code to ensure the thing wouldn’t barf if GPIO
isn’t found, to assist with development scenarios.



*From:* Mike DeVita [mailto:notifications@github.com]
*Sent:* Wednesday, February 25, 2015 12:07 PM
*To:* Dan-in-CA/OSPi
*Cc:* Jonathan Marsh
*Subject:* Re: [OSPi] using modern web technologies like bootstrap (#93)



@jonathan-marsh <https://github.com/jonathan-marsh> server actually starts
up without being able to talk to the gpio's? Wild. I know when I was
dev'ing the nodejs version of a sprinkler app it would barf if it couldn't
init the GPIO pins.

—
Reply to this email directly or view it on GitHub
<https://github.com/Dan-in-CA/OSPi/issues/93#issuecomment-76044831>.

**Comment by Dan-in-CA** [comment]
Due to dev on Win7 there are several places in the Python code that get around not having access to the GPIO pins.

**Comment by william-powell** [comment]
Running the ospi application on IE 11 has rendering issues. *GASP* I was able to mitigate by adding the following meta tag to base.html

<meta http-equiv="x-ua-compatible" content="IE=edge">

That allowed the app to render properly in IE. That tag should only affect IE. I tested in Chrome, and the web application continued to function.

Thanks for the awesome work on this project!

**Comment by Dan-in-CA** [comment]
Thanks William! I have updated the file in the repository.

**Comment by Dan-in-CA** [comment]
This issue regarding an update to the UI seems to be inactive Therefore I plan to close it in the next few days.

**Comment by Dan-in-CA** [comment]
This seems to no longer be active.

**Comment by Dan-in-CA** [comment]
Update 11/3/2016:
Bootstrap has been added to SIP and the native web UI works quite well on small screen mobile devices (Phones) with only minor modifications.

**Comment by mikedevita** [comment]
thanks @Dan-in-CA ill download the latest and try it out, ive been working on other projects and obv never got time to work on this.

**Comment by Dan-in-CA** [comment]
Cool, great to hear from you Mike.

I am still getting to know Bootstrap and have started to try and modify 
the the Timepicker bootstrap plugin. It does not handle setting just 
seconds and minutes out of the box.

At first there was a problem upgrading to the latest jQuery and I 
thought it was going to require a major overhaul of the existing UI but 
it only needed a single line "<!DOCTYPE html>" added to the "base" HTML 
template.

Once jQuery was up to date and I added bootstrap to the "static" 
directory it just started working to my total amazement. I have made a 
few tweaks but there are still some minor problems.

Dan
